File: leetcode/130-surrounded-regions.js
// https://leetcode.com/problems/surrounded-regions/ (top-interview-questions)

/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solve = function(board) {

  const m = board.length;
  const n = board[0].length;


  const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];
                  //   down      up     right    left
  
  const dfs = (r, c) => {
    if (r < 0 || r >= m || c < 0 || c >= n || board[r][c] !== 'O') return;
    board[r][c] = 'U';
    for (const direction of directions) {
      dfs(r+direction[0], c+direction[1]);
    }
  }

  // mark the not surrounded regions with U insted of O...
  for (let r = 0; r < m; r += 1) {
    for (let c = 0; c < n; c += 1) {
      if (r !== 0 && r !== (m-1) && c > 0 && c < (n-1)) continue;
      if (board[r][c] === 'O') {
        // spread around this 'O's region and mark it as unsurrounded
        dfs(r, c);
      }
    }
  }

  for (let r = 0; r < m; r += 1) {
    for (let c = 0; c < n; c += 1) {
      if (board[r][c] === 'O') board[r][c] = 'X';
      else if (board[r][c] === 'U') board[r][c] = 'O';
    }
  }

};

File: leetcode/130-surrounded-regions.py
# https://leetcode.com/problems/surrounded-regions/ (top-interview-questions)

from collections import deque
from typing import List


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        def markAsUnsurroundedDFS(row, col):
            if min(row, col) < 0 or row >= len(board) or col >= len(board[0]) or board[row][col] != "O":
                return
            board[row][col] = "U"
            markAsUnsurroundedDFS(row+1, col)
            markAsUnsurroundedDFS(row-1, col)
            markAsUnsurroundedDFS(row, col+1)
            markAsUnsurroundedDFS(row, col-1)

        def markAsUnsurroundedBFS(row, col):
            board[row][col] = "U"
            queue = deque()
            queue.append((row, col))
            directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
            while queue:
                r, c = queue.popleft()
                for dr, dc in directions:
                    if min(r+dr, c+dc) >= 0 and r+dr < len(board) and c+dc < len(board[0]) and board[r+dr][c+dc] == "O":
                        board[r+dr][c+dc] = "U"
                        queue.append((r+dr, c+dc))

        for i in range(len(board)):
            if board[i][0] == "O":
                markAsUnsurroundedBFS(i, 0)
            if board[i][-1] == "O":
                markAsUnsurroundedBFS(i, len(board[0])-1)

        for i in range(len(board[0])):
            if board[0][i] == "O":
                markAsUnsurroundedBFS(0, i)
            if board[-1][i] == "O":
                markAsUnsurroundedBFS(len(board)-1, i)
        
        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == "O":
                    board[r][c] = "X"
                elif board[r][c] == "U":
                    board[r][c] = "O"
